File: logdemo/src/main/java/ua/at/tsvetkov/demo/MainActivity.java
package ua.at.tsvetkov.demo;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.view.View;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

import ua.at.tsvetkov.util.Log;
import ua.at.tsvetkov.util.LongLog;
import ua.at.tsvetkov.util.interceptor.LogToFileInterceptor;

public class MainActivity extends AppCompatActivity {

   private static final String MSG = "TEST MESSAGE\nLINE 1\nLINE 2\nLINE 3";
   private static final String SHORT_MSG = "TEST MESSAGE";
   private static final String LONG_MSG = "TEST MESSAGE\nLINE 1\nLINE 2\nLINE 3" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 1" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 2" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 3" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 4" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 5" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 6" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 7" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 8" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 9" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 10" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 11" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 12" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 13" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 14" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 15" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 16" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 17" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 18" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 19" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 20" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 21" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 22" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 23" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 24" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 25" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 26" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 27" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 28" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 29" +
           "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 30";

   @Override
   protected void onCreate(Bundle savedInstanceState) {
      super.onCreate(savedInstanceState);
      setContentView(R.layout.activity_main);

      final LogToFileInterceptor interceptor = new LogToFileInterceptor( this);

      Log.addInterceptor(interceptor);

      runLogExamples();

      runFragmentStackLoggerExample();

      runLongLogExamples();

      Test t = new Test("TTTTTT");
      t.getName();

      findViewById(R.id.button).setOnClickListener(new View.OnClickListener() {
         @Override
         public void onClick(View v) {
            interceptor.shareZippedLog(MainActivity.this);
         }
      });

   }

   private void runLogExamples() {
      // Code example

      Log.v("Verbose");
      Log.d("Debug");
      Log.i("Info");
      Log.e("Error");
      try {
         int i = 10 / 0;
      } catch (Exception e) {
         Log.e("Some exception", e);
      }

      // Logging tests

      Exception ex = new Exception("Boo Exception");

      Log.v(ex);
      Log.v(MSG);
      Log.v(this, MSG);
      Log.v(MSG, ex);
      Log.v(this, MSG, ex);

      Log.d(ex);
      Log.d(MSG);
      Log.d(this, MSG);
      Log.d(MSG, ex);
      Log.d(this, MSG, ex);

      Log.i(ex);
      Log.i(MSG);
      Log.i(this, MSG);
      Log.i(MSG, ex);
      Log.i(this, MSG, ex);

      Log.w(ex);
      Log.w(MSG);
      Log.w(this, MSG);
      Log.w(MSG, ex);
      Log.w(this, MSG, ex);

      Log.e(ex);
      Log.e(MSG);
      Log.e(this, MSG);
      Log.e(MSG, ex);
      Log.e(this, MSG, ex);

      Log.wtf(ex);
      Log.wtf(MSG);
      Log.wtf(this, MSG);
      Log.wtf(MSG, ex);
      Log.wtf(this, MSG, ex);

      Thread th = new Thread("Boo Thread");
      th.start();

      Log.threadInfo();
      Log.threadInfo(MSG);
      Log.threadInfo(ex);
      Log.threadInfo(MSG, ex);
      Log.threadInfo(th, ex);

      Log.stackTrace();
      Log.stackTrace(MSG);

      RuntimeException rt = new RuntimeException();
      try {
         Log.rt(rt);
      } catch (RuntimeException e) {
         Log.v("Intercepted", e);
      }
      try {
         Log.rt(MSG, rt);
      } catch (RuntimeException e) {
         Log.v("Intercepted", e);
      }

      int[] ints = new int[]{472834, 4235, 657, -1728, 0};
      Log.array(ints);
      double[] doubles = new double[]{472834, 4235, 657, -1728, 0};
      Log.array(doubles);
      long[] longs = new long[]{472834, 4235, 657, -1728, 0};
      Log.array(longs);
      float[] floats = new float[]{645.0f, 235, 57, -128, 0};
      Log.array(floats);
      boolean[] bools = new boolean[]{true, false, true};
      Log.array(bools);
      char[] chars = new char[]{'c', 'h', 'a', 'r', 's'};
      Log.array(chars);
      Object[] objects = new Object[]{"String object", new Object(), 'a', new Object(), 's'};
      Log.array(objects);

      Map<Integer, String> map = new HashMap<>();
      map.put(1, "First");
      map.put(2, "Second");
      map.put(3, "Fird");
      Log.map(map);

      ArrayList<String> list = new ArrayList<>();
      list.add("First");
      list.add("Second");
      list.add("Fird");
      Log.list(list);

      Log.objl(list);
      Log.objn(map);

      byte[] data = new byte[]{5, 65, 23, 34, 32, 12, 45, 35, 66, 120, 100, 93, 31, 111};
      Log.hex(data);
      Log.hex(data, 5);

      String xml = "<note>\n" +
              "<to>Tove</to>\n" +
              "<from>Jani</from>\n" +
              "<heading>Reminder</heading>\n" +
              "<body>Don't forget me this weekend!</body>\n" +
              "</note>";
      Log.xml(xml);
      Log.xml(xml, 3);

   }

   private void runLongLogExamples() {
      // Code example

      // LongLogging tests

      Random random = new Random();

      LongLog.v(LONG_MSG);
      try {
         int i = 10 / 0;
      } catch (Exception e) {
         LongLog.e(LONG_MSG, e);
      }

      Exception ex = new Exception(LONG_MSG);

      LongLog.v(LONG_MSG);
      LongLog.v(this, LONG_MSG);
      LongLog.v(LONG_MSG, ex);
      LongLog.v(this, LONG_MSG, ex);

      LongLog.d(LONG_MSG);
      LongLog.d(this, LONG_MSG);
      LongLog.d(LONG_MSG, ex);
      LongLog.d(this, LONG_MSG, ex);

      LongLog.i(LONG_MSG);
      LongLog.i(this, LONG_MSG);
      LongLog.i(LONG_MSG, ex);
      LongLog.i(this, LONG_MSG, ex);

      LongLog.w(LONG_MSG);
      LongLog.w(this, LONG_MSG);
      LongLog.w(LONG_MSG, ex);
      LongLog.w(this, LONG_MSG, ex);

      LongLog.e(LONG_MSG);
      LongLog.e(this, LONG_MSG);
      LongLog.e(LONG_MSG, ex);
      LongLog.e(this, LONG_MSG, ex);

      LongLog.wtf(LONG_MSG);
      LongLog.wtf(this, LONG_MSG);
      LongLog.wtf(LONG_MSG, ex);
      LongLog.wtf(this, LONG_MSG, ex);

      Thread th = new Thread(LONG_MSG);
      th.start();

      LongLog.threadInfo();
      LongLog.threadInfo(LONG_MSG);
      LongLog.threadInfo(ex);
      LongLog.threadInfo(LONG_MSG, ex);
      LongLog.threadInfo(th, ex);

      LongLog.stackTrace();
      LongLog.stackTrace(LONG_MSG);

      RuntimeException rt = new RuntimeException();
      try {
         LongLog.rt(LONG_MSG, rt);
      } catch (RuntimeException e) {
         LongLog.v("Intercepted", e);
      }

      int[] ints = new int[5000];
      for (int i = 0; i < ints.length; i++) {
         ints[i] = random.nextInt();
      }
      LongLog.array(ints);

      double[] doubles = new double[5000];
      for (int i = 0; i < doubles.length; i++) {
         doubles[i] = random.nextDouble();
      }
      LongLog.array(doubles);

      long[] longs = new long[5000];
      for (int i = 0; i < longs.length; i++) {
         longs[i] = random.nextLong();
      }
      LongLog.array(longs);

      float[] floats = new float[5000];
      for (int i = 0; i < floats.length; i++) {
         floats[i] = random.nextFloat();
      }
      LongLog.array(floats);

      boolean[] bools = new boolean[5000];
      for (int i = 0; i < bools.length; i++) {
         bools[i] = random.nextBoolean();
      }
      LongLog.array(bools);
      char[] chars = new char[5000];
      for (int i = 0; i < chars.length; i++) {
         chars[i] = (char) random.nextInt(30);
      }
      LongLog.array(chars);

      Object[] objects = new Object[1000];
      for (int i = 0; i < objects.length; i++) {
         objects[i] = new Object();
      }
      LongLog.array(objects);

      Map<Integer, String> map = new HashMap<>();
      for (int i = 0; i < 500; i++) {
         map.put(i, SHORT_MSG);
      }
      LongLog.map(map);

      ArrayList<String> list = new ArrayList<>();
      for (int i = 0; i < 500; i++) {
         list.add(SHORT_MSG);
      }
      LongLog.list(list);

      LongLog.objl(list);
      LongLog.objn(list);

      byte[] data = new byte[5000];
      for (int i = 0; i < data.length; i++) {
         data[i] = (byte) random.nextInt(60);
      }
      LongLog.hex(data);
      LongLog.hex(data, 20);

      StringBuilder sb = new StringBuilder("<note>\n");
      for (int i = 0; i < 500; i++) {
         sb.append("<to" + i + ">Tove" + i + "</to" + i + ">\n");
      }
      sb.append("</note>");
      String xml = sb.toString();
      LongLog.xml(xml);
      LongLog.xml(xml, 3);

   }

   private void runFragmentStackLoggerExample() {
      getSupportFragmentManager().beginTransaction()
              .replace(R.id.content, new Fragment())
              .addToBackStack("Fragment New")
              .commit();

      new Thread(new Runnable() {
         @Override
         public void run() {
            try {
               Thread.sleep(1000);
            } catch (InterruptedException e) {
               Log.e(e);
            }
            replaceFragment();
         }
      }).start();
   }

   private void replaceFragment() {
      runOnUiThread(new Runnable() {
         @Override
         public void run() {
            TestFragment fr = new TestFragment();
            getSupportFragmentManager().beginTransaction()
                    .replace(R.id.content, fr)
                    .addToBackStack("TestFragment")
                    .commit();
            popFragment();
         }
      });
   }

   private void popFragment() {
      new Thread(new Runnable() {
         @Override
         public void run() {
            try {
               Thread.sleep(1000);
            } catch (InterruptedException e) {
               Log.e(e);
            }
            getSupportFragmentManager().popBackStack();
         }
      }).start();
   }

}

File: demo/src/main/java/ua/at/tsvetkov/demo/MainActivity.kt
/******************************************************************************
 * Copyright (c) 2010-2025 Alexandr Tsvetkov.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the MIT License
 * which accompanies this distribution, and is available at
 * https://opensource.org/license/mit
 *
 * * Contributors:
 * Alexandr Tsvetkov - initial API and implementation
 *
 * Project:
 * TAO Log
 *
 * License agreement:
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
 * associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute,
 * sublicense, and/or sell copies of the Software, and to permit persons to whom the Software
 * is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package ua.at.tsvetkov.demo

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GravityCompat
import com.google.android.material.navigation.NavigationView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import ua.at.tsvetkov.demo.databinding.ActivityMainBinding
import ua.at.tsvetkov.util.logger.Log
import ua.at.tsvetkov.util.logger.LogLong
import ua.at.tsvetkov.util.logger.PacketLog
import ua.at.tsvetkov.util.logger.interceptor.Level
import ua.at.tsvetkov.util.logger.ui.LogFragment
import java.util.*

/**
 * Created by Alexandr Tsvetkov on 06.12.2017.
 */
class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {


    private lateinit var binding: ActivityMainBinding

    private val testObject = TestObject("A Test Object")
    private lateinit var toggle: ActionBarDrawerToggle
    private lateinit var logFragment: LogFragment
    private lateinit var testFragment1: TestFragment1
    private lateinit var testFragment2: TestFragment2
    private lateinit var testFragment3: TestFragment3


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.appBarMain.toolbar)
        toggle = ActionBarDrawerToggle(
            this,
            binding.drawerLayout,
            binding.appBarMain.toolbar,
            R.string.navigation_drawer_open,
            R.string.navigation_drawer_close
        )
        binding.drawerLayout.addDrawerListener(toggle)
        toggle.syncState()
        binding.navView.setNavigationItemSelectedListener(this)
        logFragment = LogFragment()
        testFragment1 = TestFragment1()
        testFragment2 = TestFragment2()
        testFragment3 = TestFragment3()

        supportFragmentManager
            .beginTransaction()
            .add(R.id.frameContent, logFragment)
            .commit()

        val callback = object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                if (binding.drawerLayout.isDrawerOpen(GravityCompat.START)) {
                    binding.drawerLayout.closeDrawer(GravityCompat.START)
                } else {
                    isEnabled = false
                    onBackPressedDispatcher.onBackPressed()
                    isEnabled = true
                }
            }
        }
        onBackPressedDispatcher.addCallback(this, callback)
    }

    @SuppressLint("NonConstantResourceId")
    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        if (binding.drawerLayout.isDrawerOpen(GravityCompat.START)) {
            binding.drawerLayout.closeDrawer(GravityCompat.START)
        }
        runScope(item)
        return true
    }

    private fun runScope(item: MenuItem) {
        when (item.itemId) {
            R.id.nav_log_quick -> runLogQuick()
            R.id.nav_log_common -> runLogCommons()
            R.id.nav_log_thread -> runLogThread()
            R.id.nav_log_arrays -> runLogArrays()
            R.id.nav_log_xml_hex -> runLogXmlHex()
            R.id.nav_log_objects -> runLogClassAndObjects()
            R.id.nav_log_long_common -> runLogLongCommons()
            R.id.nav_log_long_thread -> runLogLongThread()
            R.id.nav_log_long_arrays -> runLogLongArrays()
            R.id.nav_log_long_xml_hex -> runLongLogXmlHex()
            R.id.nav_log_packet_log -> runPacketLog()
            R.id.nav_log_component_activity -> runActivity()
            R.id.nav_log_component_fragment -> runFragment()
        }
        logFragment.scrollDown()
    }

    private fun runLogQuick() {
        Log.v(Level.VERBOSE.name)
        Log.d(Level.DEBUG.name)
        Log.i(Level.INFO.name)
        Log.w(Level.WARNING.name)
        Log.e(Level.ERROR.name)
        Log.wtf(Level.WTF.name)
    }

    private fun runLogCommons() {
        try {
           val i = 10 / 0
        } catch (e: Exception) {
            Log.e("Exception", e)
        }
        val ex = Exception("Log Exception")
        Log.v(ex)
        Log.v(MSG)
        Log.v(MSG, ex)
        Log.d(ex)
        Log.d(MSG)
        Log.d(MSG, ex)
        Log.i(ex)
        Log.i(MSG)
        Log.i(MSG, ex)
        Log.w(ex)
        Log.w(MSG)
        Log.w(MSG, ex)
        Log.e(ex)
        Log.e(MSG)
        Log.e(MSG, ex)
        Log.wtf(ex)
        Log.wtf(MSG)
        Log.wtf(MSG, ex)
        val rt = RuntimeException()
        try {
            Log.rt(rt)
        } catch (e: RuntimeException) {
            Log.v("Intercepted", e)
        }
        try {
            Log.rt(MSG, rt)
        } catch (e: RuntimeException) {
            Log.v("Intercepted", e)
        }
    }

    private fun runLogThread() {
        val th = Thread("Test Thread")
        th.start()
        val ex = Exception("Thread Exception")
        Log.threadInfo()
        Log.threadInfo(MSG)
        Log.threadInfo(ex)
        Log.threadInfo(MSG, ex)
        Log.threadInfo(th, ex)
        Log.stackTrace()
        Log.stackTraceV(MSG)
        Log.stackTraceI(MSG)
        Log.stackTraceD(MSG)
        Log.stackTraceW(MSG)
        Log.stackTraceE(MSG)
        Log.stackTraceWTF(MSG)
    }

    private fun runLogArrays() {
        val ints = intArrayOf(472834, 4235, 657, -1728, 0)
        Log.array(ints)
        val doubles = doubleArrayOf(472834.0, 4235.0, 657.0, -1728.0, 0.0)
        Log.array(doubles)
        val longs = longArrayOf(472834, 4235, 657, -1728, 0)
        Log.array(longs)
        val floats = floatArrayOf(645.0f, 235f, 57f, -128f, 0f)
        Log.array(floats)
        val bools = booleanArrayOf(true, false, true)
        Log.array(bools)
        val chars = charArrayOf('c', 'h', 'a', 'r', 's')
        Log.array(chars)
        val objects = arrayOf("String object", Any(), 'a', Any(), 's')
        Log.array(objects)
        val map: MutableMap<Int?, String?> = HashMap()
        map[1] = "First"
        map[2] = "Second"
        map[3] = "Third"
        Log.map(map)
        val list = ArrayList<String?>()
        list.add("First")
        list.add("Second")
        list.add("Third")
        Log.list(list)
    }

    private fun runLogXmlHex() {
        val data = byteArrayOf(5, 65, 23, 34, 32, 12, 45, 35, 66, 120, 100, 93, 31, 111)
        Log.hex(data)
        Log.hex(data, 5)
        val xml = """
            <note>
            <to>Alex</to>
            <from>July</from>
            <heading>Reminder</heading>
            <body>Don't forget me this weekend!</body>
            </note>
            """.trimIndent()
        Log.xml(xml)
        Log.xml(xml, 3)
    }

    private fun runLogClassAndObjects() {
        Log.classInfo(testObject)
        Log.objectInfo(testObject)
    }

    private fun runLogLongCommons() {
        val ex = Exception()
        LogLong.v(LONG_MSG)
        LogLong.v(LONG_MSG, ex)
        LogLong.d(LONG_MSG)
        LogLong.d(LONG_MSG, ex)
        LogLong.i(LONG_MSG)
        LogLong.i(LONG_MSG, ex)
        LogLong.w(LONG_MSG)
        LogLong.w(LONG_MSG, ex)
        LogLong.e(LONG_MSG)
        LogLong.e(LONG_MSG, ex)
        LogLong.wtf(LONG_MSG)
        LogLong.wtf(LONG_MSG, ex)
        val rt = RuntimeException()
        try {
            LogLong.rt(LONG_MSG, rt)
        } catch (e: RuntimeException) {
            LogLong.v("Intercepted", e)
        }
    }

    private fun runLogLongThread() {
        val ex = Exception()
        val th = Thread(LONG_MSG)
        th.start()
        LogLong.threadInfo()
        LogLong.threadInfo(LONG_MSG)
        LogLong.threadInfo(ex)
        LogLong.threadInfo(LONG_MSG, ex)
        LogLong.threadInfo(th, ex)
        LogLong.stackTrace()
        LogLong.stackTraceV(LONG_MSG)
        LogLong.stackTraceI(LONG_MSG)
        LogLong.stackTraceD(LONG_MSG)
        LogLong.stackTraceW(LONG_MSG)
        LogLong.stackTraceE(LONG_MSG)
        LogLong.stackTraceWTF(LONG_MSG)
    }

    private fun runLogLongArrays() {
        val random = Random()
        val ints = IntArray(5000)
        for (i in ints.indices) {
            ints[i] = random.nextInt()
        }
        LogLong.array(ints)
        val doubles = DoubleArray(5000)
        for (i in doubles.indices) {
            doubles[i] = random.nextDouble()
        }
        LogLong.array(doubles)
        val longs = LongArray(5000)
        for (i in longs.indices) {
            longs[i] = random.nextLong()
        }
        LogLong.array(longs)
        val floats = FloatArray(5000)
        for (i in floats.indices) {
            floats[i] = random.nextFloat()
        }
        LogLong.array(floats)
        val bools = BooleanArray(5000)
        for (i in bools.indices) {
            bools[i] = random.nextBoolean()
        }
        LogLong.array(bools)
        val chars = CharArray(5000)
        for (i in chars.indices) {
            chars[i] = random.nextInt(30).toChar()
        }
        LogLong.array(chars)
        val objects = arrayOf<Any>(100)
        for (i in objects.indices) {
            objects[i] = Any()
        }
        LogLong.array(objects)
        val map: MutableMap<Int, String> = HashMap()
        for (i in 0..499) {
            map[i] = SHORT_MSG
        }
        LogLong.map(map)
        val list = ArrayList<String>()
        for (i in 0..499) {
            list.add(SHORT_MSG)
        }
        LogLong.list(list)
    }

    private fun runLongLogXmlHex() {
        val random = Random()
        val data = ByteArray(5000)
        for (i in data.indices) {
            data[i] = random.nextInt(60).toByte()
        }
        LogLong.hex(data)
        LogLong.hex(data, 20)
        val sb = StringBuilder("<note>\n")
        for (i in 0..499) {
            sb
                .append("<to")
                .append(i)
                .append(">Tove")
                .append(i)
                .append("</to")
                .append(i)
                .append(">\n")
        }
        sb.append("</note>")
        val xml = sb.toString()
        LogLong.xml(xml)
        LogLong.xml(xml, 3)
    }

    private fun runPacketLog() {
        val packet = PacketLog()
        packet.logI("PacketLog test started")
        packet.logD("This is a debug message")
        packet.logW("This is a warning")
        packet.logE("This is an error")
        packet.logV("And a verbose message")
        packet.printLogs("PacketLog Test")
    }

    private fun runActivity() {
        val intent = Intent(this, TestActivity::class.java)
        startActivity(intent)
    }

    private fun runFragment() {
        supportFragmentManager.beginTransaction()
            .replace(R.id.frameContent, testFragment1)
            .addToBackStack("TestFragment1")
            .commit()

        CoroutineScope(Dispatchers.Default).launch {
            delay(500)
            withContext(Dispatchers.Main) {
                supportFragmentManager.beginTransaction()
                    .replace(R.id.frameContent, testFragment2)
                    .addToBackStack("TestFragment2")
                    .commit()
            }
            delay(500)
            withContext(Dispatchers.Main) {
                supportFragmentManager.beginTransaction()
                    .replace(R.id.frameContent, testFragment3)
                    .addToBackStack("TestFragment3")
                    .commit()
            }
            Log.i("You can see changes in the fragment stack in your LogCat")
        }
    }

    companion object {
        private const val MSG = "TEST MESSAGE\nLINE 1\nLINE 2\nLINE 3"
        private const val SHORT_MSG = "TEST MESSAGE"
        private const val LONG_MSG = "TEST MESSAGE\nLINE 1\nLINE 2\nLINE 3" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 1" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 2" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 3" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 4" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 5" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 6" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 7" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 8" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 9" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 10" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 11" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 12" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 13" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 14" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 15" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 16" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 17" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 18" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 19" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 20" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 21" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 22" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 23" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 24" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 25" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 26" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 27" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 28" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 29" +
                "\n a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message a message 30"
    }
}
